### PR TITLE
WAGE: Various bug fixes and Improvements

### PR DESCRIPTION
**Pull request by StableSteady**

### CODE REVIEW

File: engines/wage/design.cpp [hunk 1]
```diff
@@ -411,8 +479,15 @@ void Design::drawOval(Graphics::ManagedSurface *surface, Common::ReadStream &in,
...
+	if (borderThickness > 1) {
+		x1 += borderThickness / 2;
+		y1 += borderThickness / 2;
+		x2 -= (borderThickness - 1) / 2;
+		y2 -= (borderThickness - 1) / 2;
+	}
+
 	if (borderThickness > 0 && borderFillType <= patterns.size())
-		Graphics::drawEllipse(x1, y1, x2-1, y2-1, kColorBlack, false, drawPixel, &pd);
+		Graphics::drawEllipse(x1, y1, x2-1, y2-1, kColorBlack, false, drawPixelCircle, &pd);
```
Comment (sev-) [inline engines/wage/design.cpp:490]: Please write `x2 - 1, y2 - 1`. E.g. spaces around operators

File: engines/wage/entities.cpp [hunk 2]
```diff
@@ -159,6 +159,13 @@ void Scene::paint(Graphics::ManagedSurface *surface, int x, int y) {
 Designed *Scene::lookUpEntity(int x, int y) {
 	for (ObjList::const_iterator it = _objs.end(); it != _objs.begin(); ) {
 		it--;
+		if (((*it)->_name == "continue" && (*it)->_resourceId == 22259) ||
```
Comment (sev-) [inline engines/wage/entities.cpp:162]: Please add a comment explaining the reason here.

File: engines/wage/entities.cpp [hunk 3]
```diff
@@ -159,6 +159,13 @@ void Scene::paint(Graphics::ManagedSurface *surface, int x, int y) {
...
+			((*it)->_name == "Goth" && (*it)->_resourceId == 18623))
+			if (_design->isPointOpaque(x, y))
+				return *it;
+
+		if ((*it)->_name == "TWIST.CLICK")
```
Comment (sev-) [inline engines/wage/entities.cpp:167]: Same thing, please explain in the comment what does it do.

Alll these comments need to start from "WORKAROUND: "

File: engines/wage/metaengine.cpp [hunk 4]
```diff
@@ -30,6 +30,10 @@
 
 namespace Wage {
 
+void WageEngine::initFeatures() {
```
Comment (sev-) [inline engines/wage/metaengine.cpp:33]: The conventional way is implementing

```C
uint32 WageEngine::getFeatures() {
   return gameDescription->flags;
}
```

and then use this function everywhere.

File: engines/wage/wage.h [hunk 5]
```diff
@@ -103,6 +104,11 @@ enum {
 	// the current limitation is 32 debug levels (1 << 31 is the last one)
 };
 
+enum Resolution {
+	GF_RES800 = 1 << 0,
```
Comment (sev-) [inline engines/wage/wage.h:108]: Please align these vertically

File: engines/wage/world.cpp [hunk 6]
```diff
@@ -215,6 +215,9 @@ bool World::loadWorld(Common::MacResManager *resMan) {
 		if (res != NULL) {
 			scene->_textBounds = readRect(res);
 			int fontType = res->readUint16BE();
+			if (_name == "***DUNE ETERNITY*** ")
```
Comment (sev-) [inline engines/wage/world.cpp:218]: Please add explanation in a comment why is this needed

File: engines/wage/util.cpp [hunk 7]
```diff
@@ -53,10 +53,10 @@ namespace Wage {
 Common::Rect *readRect(Common::SeekableReadStream *in) {
 	int x1, y1, x2, y2;
 
-	y1 = in->readSint16BE();
-	x1 = in->readSint16BE();
-	y2 = in->readSint16BE() + 4;
-	x2 = in->readSint16BE() + 4;
+	y1 = in->readSint16BE() - 2;
```
Comment (sev-) [inline engines/wage/util.cpp:56]: Please explain in a comment why we're doing this

File: graphics/macgui/macdialog.cpp [hunk 8]
```diff
@@ -44,36 +44,40 @@
  *
  */
 
-#include "common/system.h"
 #include "common/events.h"
+#include "common/system.h"
+
+//#include "engines/engine.h"
```
Comment (sev-) [inline graphics/macgui/macdialog.cpp:50]: Maybe remove it altogether?

File: graphics/macgui/macdialog.cpp [hunk 9]
```diff
@@ -116,9 +119,9 @@ void Dialog::paint() {
 
 		if ((int)i == _pressedButton && _mouseOverPressedButton) {
 			Common::Rect bb(button->bounds.left + 5, button->bounds.top + 5,
-				button->bounds.right - 5, button->bounds.bottom - 5);
+							button->bounds.right - 5, button->bounds.bottom - 5);
 
-			Design::drawFilledRect(&_gui->_screen, bb, kColorBlack, _gui->_wm->getPatterns(), kPatternSolid);
+			Wage::Design::drawFilledRect(&_gui->_screen, bb, kColorBlack, _gui->_wm->getPatterns(), kPatternSolid);
```
Comment (sev-) [inline graphics/macgui/macdialog.cpp:124]: Err, how Graphics:: could call a specific engine? Please refactor, the common code cannot refer to any engine

File: graphics/macgui/macdialog.cpp [hunk 10]
```diff
@@ -44,36 +44,40 @@
  *
  */
 
-#include "common/system.h"
 #include "common/events.h"
+#include "common/system.h"
+
+//#include "engines/engine.h"
+#include "engines/wage/design.h"
```
Comment (sev-) [inline graphics/macgui/macdialog.cpp:51]: This should never be included here.

**Review by sev- — CHANGES_REQUESTED** [review]
Needs some changes. The major one is the sudden dependency of graphics/macgui code from engines/wage. This must be untangled.

File: engines/wage/design.cpp [hunk 11]
```diff
@@ -372,6 +433,13 @@ void Design::drawPolygon(Graphics::ManagedSurface *surface, Common::ReadStream &
 	xcoords.push_back(x1);
 	ycoords.push_back(y1);
 
+	if (borderThickness > 1) {
+		for (int i = 0; i < xcoords.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (StableSteady) [reply engines/wage/design.cpp]: Done

File: engines/wage/design.cpp [hunk 12]
```diff
@@ -253,6 +253,53 @@ void drawPixel(int x, int y, int color, void *data) {
...
+				for (x = x1; x < x2; x++)
+					p->design->adjustBounds(x, y);
+		}
+
+		return;
+	}
+
+	byte *pat = p->patterns->operator[](p->fillType - 1);
+
+	if (p->thickness == 1) {
```
[the remark this answers is not in the document]
Comment (StableSteady) [reply engines/wage/design.cpp:284]: Done

File: engines/wage/design.cpp [hunk 13]
```diff
@@ -253,6 +253,53 @@ void drawPixel(int x, int y, int color, void *data) {
...
+		if (x >= 0 && x < p->surface->w && y >= 0 && y < p->surface->h) {
+			uint xu = (uint)x; // for letting compiler optimize it
+			uint yu = (uint)y;
+
+			*((byte *)p->surface->getBasePtr(xu, yu)) =
+				(pat[yu % 8] & (1 << (7 - xu % 8))) ? color : kColorWhite;
+		}
+	} else {
+		int x1 = x - p->thickness / 2;
```
[the remark this answers is not in the document]
Comment (StableSteady) [reply engines/wage/design.cpp:293]: My test already had all possible thickness and this behavior works for them.

File: engines/wage/design.cpp (hunk 1 above)
Comment (StableSteady) [reply engines/wage/design.cpp]: Done

File: engines/wage/entities.cpp (hunk 2 above)
Comment (StableSteady) [reply engines/wage/entities.cpp:163]: Done

File: engines/wage/entities.cpp (hunk 3 above)
Comment (StableSteady) [reply engines/wage/entities.cpp:169]: Done

File: engines/wage/metaengine.cpp (hunk 4 above)
Comment (StableSteady) [reply engines/wage/metaengine.cpp]: Done

File: engines/wage/wage.h (hunk 5 above)
Comment (StableSteady) [reply engines/wage/wage.h]: Done

File: engines/wage/world.cpp (hunk 6 above)
Comment (StableSteady) [reply engines/wage/world.cpp:219]: Done

File: engines/wage/util.cpp (hunk 7 above)
Comment (StableSteady) [reply engines/wage/util.cpp:57]: Done

File: engines/wage/world.cpp [hunk 14]
```diff
@@ -215,6 +215,10 @@ bool World::loadWorld(Common::MacResManager *resMan) {
 		if (res != NULL) {
 			scene->_textBounds = readRect(res);
 			int fontType = res->readUint16BE();
+			// WORKAROUND: Dune Eternity has a weird fontType ID so we override it to the correct one
```
Comment (sev-) [inline engines/wage/world.cpp:218]: Could you please specify what is this "weird type ID"

File: graphics/macgui/macdialog.cpp [hunk 15]
```diff
@@ -116,97 +115,61 @@ void Dialog::paint() {
...
-void Dialog::drawOutline(Common::Rect &bounds, int *spec, int speclen) {
+void MacDialog::drawOutline(Common::Rect &bounds, int *spec, int speclen) {
+	MacPlotData pd(_screen, nullptr, &_wm->getPatterns(), 1, 0, 0, 1, _wm->_colorBlack, false);	
 	for (int i = 0; i < speclen; i++)
-		if (spec[i] != 0)
-			Design::drawRect(&_gui->_screen, bounds.left + i, bounds.top + i, bounds.right - i, bounds.bottom - i,
-						1, kColorBlack, _gui->_wm->getPatterns(), kPatternSolid);
-}
-
-int Dialog::run() {
```
Comment (sev-) [inline graphics/macgui/macdialog.cpp:147]: I wonder, why is this running on an engine level but is not here? You may just return different things like "should quit" if EVEN_QUIT is set and kDialogOption1/kDialogOption2 for the requested options

**Review by sev- — CHANGES_REQUESTED** [review]
Getting there. Added couple more notes

File: engines/wage/world.cpp (hunk 14 above)
Comment (StableSteady) [reply engines/wage/world.cpp:218]: The fontType it reads is 14614 which is way outside of the range of font Types supported. So I hardcoded it to what it should be.

File: engines/wage/design.cpp [hunk 16]
```diff
@@ -194,6 +194,24 @@ bool Design::isPointOpaque(int x, int y) {
 	return pixel != kColorGreen;
 }
 
+bool Design::isInBounds(int x, int y) {
+	if (_surface == NULL)
+		error("Surface is null");
```
[the remark this answers is not in the document]
Comment (StableSteady) [reply engines/wage/design.cpp]: Done

**Comment by sev-** [comment]
Thanks!
